Update to cardano-node 1.35-rc2

Anviking opened the pull request:
### Overview

- Updates dependencies to target `cardano-node 1.35-rc2`
- Adds support to make the wallet _mostly_ work in Babbage
- Rework of nix setup: we now derive nix files from `cabal.project` rather than the weird `stack.yaml`-based approach we had before.


### Detailed todo-list

- [x] Get nix happy with the new dependencies
- [x] Make `cabal build all` work with modulo warnings
- [x] Update to more recent revisions
    - [x] Fix compilation in tests again
- [x] Make `cabal build all` work with _no_ warnings
    - [x] Fix puzzling persistent TH name shadowing warnings -> fixed by sam?
- [x] Ensure existing tests pass
    - [x] unit
    - [x] Look closer at balanceTx golden unit failure after first seeing if balanceTx integration tests pass on both alonzo and Babbage
    - [x] Integration (on alonzo like on master)
        - [x] Was blocked by readTVar, fixed with https://github.com/input-output-hk/io-sim/pull/6#pullrequestreview-994757891
        - [x] Fix failures relating to Babbage era -> fixed with a hack https://github.com/input-output-hk/cardano-wallet/pull/3280/commits/9e2568e8058e2e4b995363805f4381ce9976108c
        - [x] Fix cardano-wallet-core Aeson failures
- [x] More closely review code-changes
- [x] Clean up git commit history
    - [x] Flatten merge commits
    - [x] Squash back-and-forth nix and revision changes
    - [x] Rebase off master 
- [x] Ensure revisions in stack.yaml and cabal.project are updated consistently 
    - Alternatively remove `stack.yaml` and ensure ./nix/regenerate.sh still works and that we know how to update sha256map.nix — #3328 but without the sha256map.nix
- [x] Ensure CI is completely green: [Hydra jobset](https://hydra.iohk.io/jobset/Cardano/cardano-wallet-pr-3280), ADP-1888
    - There may be CI specific problems with nix, windows and more (there probably are)
    - [x] fix hydra wine TH error e.g. `$(TH.lift =<< TH.runIO (readFile "app/foo.txt"))`
          - seems relevant https://github.com/input-output-hk/haskell.nix/pull/1508
          - seems gone now
    - [x] `Ambiguous module name` https://hydra.iohk.io/build/15469467
        - https://input-output-rnd.slack.com/archives/C48HA39RS/p1654388106509579
        - Bumping Haskell.nix again should help?
        - Didn't seem like it worked...
        - https://input-output-rnd.slack.com/archives/CAP8NM7N0/p1654828014483969
     - [x] Remove-iserv problem https://hydra.iohk.io/build/15604435 
         - Update from Hamish https://input-output-rnd.slack.com/archives/CAP8NM7N0/p1655213059547499?thread_ts=1654828014.483969&cid=CAP8NM7N0
     - [x] Docs shell - we don't need explicit emanote input in nix config
     - [x] Profiled shell:  `recursionzmschemeszm5zi2zi2zi2zm6NC6xZZx2tI3826DrVhqI9v_DataziFunctorziFoldableziTH_zdfMakeBaseFunctorNamezuzdcmakeBaseFunctor_closure` https://hydra.iohk.io/build/15694577/nixlog/2/tail
     - [x] musl cli goldens https://hydra.iohk.io/build/15694618/nixlog/82
     - [x] `hie-bios: hie-bios: this operation requires -fexternal-interpreter
     - [x] `Ambiguous module name 'Options.Applicative':` in HLS - now we at least have an ugly `-XPackageImport` workaround.
- [x] Update node revision in readme.md

### Things we need eventually, but not strictly for this PR

> -  Update to final 1.35 tag once it is released - can be done in separate PR
> -  Run existing integration tests on both Alonzo and Babbage eras in CI
> - Depend on cardano-addresses with aeson2 support merged, tagged and released
> - Depend on a cardano-sl-x509 where aeson 2 support is merged to master
> - Consider appropriateness Tx deserialisation hack https://github.com/input-output-hk/cardano-wallet/pull/3280/commits/9e2568e8058e2e4b995363805f4381ce9976108c

Moved to jira tasks in https://input-output.atlassian.net/browse/ADP-1538

### Comments

Includes #3258, #3293

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

ADP-1744

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

sevanspowell committed:
libsodium.dll is now automatically installed

- Remove libsodium.dll from Windows post-install steps, as haskell.nix now
  automatically installs libsodium.

jonathanknowles commented on lib/core/test/unit/Cardano/Api/GenSpec.hs:
```diff
@@ -1184,6 +1195,28 @@ genTxOutDatumCoverage era datum =
+            & label "reference scripts not generated in unsupported era"
+            & counterexample ( "reference scripts were generated in unsupported "
+                               <> show era
+                             )
```
```suggestion
            & counterexample $ mconcat
                [ "reference scripts were generated in unsupported "
                , show era
                ]
```

jonathanknowles commented on lib/core/src/Network/Wai/Middleware/Logging.hs:
```diff
@@ -277,7 +279,7 @@ instance ToJSON HandlerLog where
 sanitize keys bytes = encode' $ case decode' bytes of
     Just (Object o) ->
-        Object (foldr (HM.adjust obfuscate) o keys)
+        Object (Aeson.fromHashMap (foldr (HM.adjust obfuscate) (Aeson.toHashMap o) (map Aeson.fromText keys)))
```
```suggestion
        Object
            $ Aeson.fromHashMap
            $ foldr
                (HM.adjust obfuscate)
                (Aeson.toHashMap o)
                (map Aeson.fromText keys)
```

jonathanknowles commented on lib/core/test/unit/Cardano/Api/GenSpec.hs:
```diff
@@ -379,6 +381,18 @@ genTxIxCoverage (TxIx ix) = unsignedCoverage (maxBound @Word16) "txIx" ix
+    arbitrary = genTxIx
+
+genCertIxCoverage :: Ledger.CertIx -> Property
+genCertIxCoverage (Ledger.CertIx ix) = unsignedCoverage (maxBound @Word16) "certIx" ix
```
```suggestion
genCertIxCoverage (Ledger.CertIx ix) =
    unsignedCoverage (maxBound @Word16) "certIx" ix
```

jonathanknowles commented on lib/core/test/unit/Cardano/Api/GenSpec.hs:
```diff
@@ -379,6 +381,18 @@ genTxIxCoverage (TxIx ix) = unsignedCoverage (maxBound @Word16) "txIx" ix
     arbitrary = genTxIndex
 
+genTxIxCoverage' :: Ledger.TxIx -> Property
+genTxIxCoverage' (Ledger.TxIx ix) = unsignedCoverage (maxBound @Word16) "txIx" ix
```
```suggestion
genTxIxCoverage' (Ledger.TxIx ix) =
    unsignedCoverage (maxBound @Word16) "txIx" ix
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1370,7 +1371,7 @@ instance FromJSON SchemaApiErrorCode where
         vals <- forM (fmap showConstr $ dataTypeConstrs $ dataTypeOf NoSuchWallet)
             $ \n -> do
-                (r :: Maybe Yaml.Value) <- o .:? T.pack (toSchemaName n)
+                (r :: Maybe Yaml.Value) <- o .:? Aeson.fromString (toSchemaName n)
```
```suggestion
                (r :: Maybe Yaml.Value) <-
                    o .:? Aeson.fromString (toSchemaName n)
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -19,11 +19,16 @@
 {-# LANGUAGE TypeOperators #-}
 {-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+
```
```suggestion
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1067,8 +1084,10 @@ toAlonzoPParams
+        , Alonzo._costmdls = either
+                                  (\e -> error $ "toAlonzoPParams: invalid cost models, error: " <> e)
+                                  id
+                                  (toAlonzoCostModels protocolParamCostModels)
```
```suggestion
        , Alonzo._costmdls = either
            (\e -> error $ "toAlonzoPParams: invalid cost models, error: " <> e)
            id
            (toAlonzoCostModels protocolParamCostModels)
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1090,21 +1109,24 @@ toAlonzoPParams
+      conv :: (Cardano.AnyPlutusScriptVersion, Cardano.CostModel) -> Either String (Alonzo.Language, Alonzo.CostModel)
+      conv (anySVer, cModel )= do
+        alonzoCostModel <- toAlonzoCostModel cModel (toAlonzoScriptLanguage anySVer)
+        Right (toAlonzoScriptLanguage anySVer, alonzoCostModel)
```
```suggestion
    toAlonzoCostModels m = do
        f <- mapM conv $ Map.toList m
        Right . Alonzo.CostModels $ Map.fromList f
      where
        conv
            :: (Cardano.AnyPlutusScriptVersion, Cardano.CostModel)
            -> Either String (Alonzo.Language, Alonzo.CostModel)
        conv (anySVer, cModel) = do
            alonzoCostModel <-
                toAlonzoCostModel cModel (toAlonzoScriptLanguage anySVer)
            Right (toAlonzoScriptLanguage anySVer, alonzoCostModel)
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1090,21 +1109,24 @@ toAlonzoPParams
+        alonzoCostModel <- toAlonzoCostModel cModel (toAlonzoScriptLanguage anySVer)
+        Right (toAlonzoScriptLanguage anySVer, alonzoCostModel)
+
+    toAlonzoCostModel :: Cardano.CostModel -> Alonzo.Language -> Either String Alonzo.CostModel
```
```suggestion
    toAlonzoCostModel
        :: Cardano.CostModel
        -> Alonzo.Language
        -> Either String Alonzo.CostModel
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1090,21 +1109,24 @@ toAlonzoPParams
+    toAlonzoScriptLanguage (Cardano.AnyPlutusScriptVersion Cardano.PlutusScriptV1) =
+        Alonzo.PlutusV1
+    toAlonzoScriptLanguage (Cardano.AnyPlutusScriptVersion Cardano.PlutusScriptV2) =
+        Alonzo.PlutusV2
```
```suggestion
    toAlonzoScriptLanguage :: Cardano.AnyPlutusScriptVersion -> Alonzo.Language
    toAlonzoScriptLanguage
        (Cardano.AnyPlutusScriptVersion Cardano.PlutusScriptV1) =
            Alonzo.PlutusV1
    toAlonzoScriptLanguage
        (Cardano.AnyPlutusScriptVersion Cardano.PlutusScriptV2) =
            Alonzo.PlutusV2
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1585,6 +1609,7 @@ getScriptIntegrityHash = \case
         Cardano.ShelleyBasedEraAllegra -> Nothing
         Cardano.ShelleyBasedEraMary    -> Nothing
         Cardano.ShelleyBasedEraAlonzo  -> SafeHash.originalBytes <$> scriptIntegrityHashOfAlonzoTx tx
+        Cardano.ShelleyBasedEraBabbage -> SafeHash.originalBytes <$> scriptIntegrityHashOfBabbageTx tx
```
```suggestion
        Cardano.ShelleyBasedEraBabbage ->
            SafeHash.originalBytes <$> scriptIntegrityHashOfBabbageTx tx
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1959,6 +1996,24 @@ toCardanoTxOut era = case era of
+        datumHash = Cardano.TxOutDatumNone
+        addrInEra = tina "toCardanoTxOut: malformed address"
+            [ Cardano.AddressInEra (Cardano.ShelleyAddressInEra Cardano.ShelleyBasedEraBabbage)
+                <$> deserialiseFromRawBytes AsShelleyAddress addr
```
```suggestion
            [ Cardano.AddressInEra
                (Cardano.ShelleyAddressInEra Cardano.ShelleyBasedEraBabbage)
                    <$> deserialiseFromRawBytes AsShelleyAddress addr
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -1012,11 +1014,11 @@ genConfig dir systemStart clusterEra logCfg = do
       where
         hardForks =
-            [ ("Test" <> T.pack (show hardFork) <> "AtEpoch", Yaml.Number 0)
+            [ (Aeson.fromText $ "Test" <> T.pack (show hardFork) <> "AtEpoch", Yaml.Number 0)
```
```suggestion
            [
                ( Aeson.fromText $ "Test" <> T.pack (show hardFork) <> "AtEpoch"
                , Yaml.Number 0
                )
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -596,6 +597,11 @@ newTransactionLayer networkId = TransactionLayer
+                    signTransaction networkId acctResolver policyResolver
+                    addressResolver inputResolver (body, wits)
+                    & sealedTxFromCardano'
+
```
```suggestion
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -2106,13 +2161,16 @@ unsealShelleyTx wtx = case W.cardanoTx wtx of
 shelleyBasedToCardanoEra Cardano.ShelleyBasedEraAllegra = AllegraEra
 shelleyBasedToCardanoEra Cardano.ShelleyBasedEraMary    = MaryEra
 shelleyBasedToCardanoEra Cardano.ShelleyBasedEraAlonzo  = AlonzoEra
+shelleyBasedToCardanoEra Cardano.ShelleyBasedEraBabbage  = BabbageEra
```
Suggestion: use `LambdaCase` to remove the repetition of the function name:

```hs
shelleyBasedToCardanoEra :: ShelleyBasedEra era -> CardanoEra era
shelleyBasedToCardanoEra = \case
    Cardano.ShelleyBasedEraShelley -> ShelleyEra
    Cardano.ShelleyBasedEraAllegra -> AllegraEra
    Cardano.ShelleyBasedEraMary    -> MaryEra
    Cardano.ShelleyBasedEraAlonzo  -> AlonzoEra
    Cardano.ShelleyBasedEraBabbage -> BabbageEra
```

jonathanknowles reviewed (commented):
Hi @Anviking 

Just leaving some suggestions (See comments.)

Anviking committed:
Try increasing node-to-client version

paweljakubas committed:
adjust Primitive.Types.Tx module

paweljakubas committed:
fromBabbageTxBodyAndAux - backbone

paweljakubas committed:
fromBabbageValidatedTx

paweljakubas committed:
fromBabbageTx

paweljakubas committed:
fromBabbageBlock

paweljakubas committed:
numberOfTransactionsInBlock

paweljakubas committed:
handle collateralOutput

paweljakubas committed:
shelley unit test fixes - part 1

Anviking committed:
Tweak error message in toPaymentHash

paweljakubas committed:
some more fillings

paweljakubas committed:
towards shelley core unit tests - part 1

paweljakubas committed:
reuse fromLedgerPParams from cardano-api - part 1

paweljakubas committed:
clean fromLedgerAlonzoPParams

paweljakubas committed:
reuse fromLedgerPParams from cardano-api - part 2

Anviking committed:
Bump io-sim to fix readTVar issue

paweljakubas committed:
assignScriptRedeemers refactoring

Anviking committed:
Hack: tweak deserialisation preference to make integration tests pass

Anviking committed:
Use Plutus.defaultCostModelParams in balanceTx tests

The hard-coded value was no longer correct.

The balanceTx pingPong_2 golden changed marginally. Presumably this is
merely due to some small difference in the cost model.

sevanspowell committed:
Restrict persistent to version < 2.13.3.4

- Something in this diff changed the behaviour of the persistent Template
Haskell: https://github.com/yesodweb/persistent/compare/persistent-2.13.3.2...persistent-2.13.3.4#diff-e762eafdebe681bccd9f2661dfeca2e9915cb56dd00facfb00fde457af5ddcafR6.
- Restrict persistent to < 2.13.3.4 to workaround this issue.

sevanspowell committed:
Upgrade version of cardano-addresses

- Upgrade version of cardano-addresses to resolve "unused package" warnings.

sevanspowell committed:
Make autoconf tools + libsecp256k1 available in non-Nix cabal build

paweljakubas committed:
fix genTxIx and genCertIx

paweljakubas committed:
try with genSlotNo

paweljakubas committed:
adjust ApiNetworkParameters in swagger

sevanspowell committed:
Upgrade version of cardano-addresses - Stack.yaml

jonathanknowles committed:
Eliminate implicit imports from `Cardano.Api.GenSpec`.

Replace all implicit imports with explicit imports.

https://github.com/input-output-hk/adrestia/blob/master/docs/code/Coding-Standards.md#we-use-explicit-imports-by-default-and-favor-qualified-imports-for-ambiguous-functions

Anviking committed:
Bump haskell.nix

paweljakubas committed:
LoggingSpec fix

Anviking committed:
Update node to 1.35.0-rc2

Co-authored-by: Pawel Jakubas <jakubas.pawel@gmail.com>
Co-authored-by: Johannes Lund <johannes.lund@iohk.io>

paweljakubas committed:
handle translation error

jonathanknowles committed:
Regenerate golden JSON files for all API types.

The `hspec-golden-aeson` library recently changed the seed size to 32
bits. See:

https://github.com/plow-technologies/hspec-golden-aeson/pull/14/files

As a result, many of our pre-existing golden JSON files have seeds that
no longer parse correctly, resulting in the following error when
attempting to run golden roundtrip tests:

```
uncaught exception: AesonDecodeError
AesonDecodeError "Error in $.seed: parsing Int32 failed, value is either
    floating or will cause over or underflow"
```

This commit regenerates golden JSON files for all API types.

jonathanknowles committed:
Add module `Cardano.Ledger.Credential.Safe`.

This module re-exports functionality provided by module
'Cardano.Ledger.Credential', but with a safer interface that does not
silently truncate slot numbers.

Anviking committed:
Bump haskell.nix again

Hoping this will fix these failures in hydra:

```
Setup.hs:10:1: error:
    Ambiguous module name `System.Process':
      it was found in multiple packages:
      process-1.6.13.2 process-1.6.13.2
   |
10 | import System.Process
   | ^^^^^^^^^^^^^^^^^^^^^

Setup.hs:11:1: error:
    Ambiguous module name `System.Directory':
      it was found in multiple packages:
      directory-1.3.6.0 directory-1.3.6.0
   |
11 | import System.Directory
   | ^^^^^^^^^^^^^^^^^^^^^^^

Setup.hs:12:1: error:
    Ambiguous module name `System.FilePath':
      it was found in multiple packages:
      filepath-1.4.2.1 filepath-1.4.2.1
   |
12 | import System.FilePath
   | ^^^^^^^^^^^^^^^^^^^^^^
```

jonathanknowles committed:
Add `genSlotNo32` and use it within `genPtr`.

Additionally, use the safe interface for `Ptr` construction provided by
`Cardano.Ledger.Credential.Safe`.

jonathanknowles committed:
Add `genSlotNo32Coverage` and use it in `genPtrCoverage`.

Additionally, use the safe interface for `Ptr` deconstruction provided
by `Cardano.Ledger.Credential.Safe`.

jonathanknowles committed:
Miscellaneous import fixes.

jonathanknowles committed:
Add roundtrip test for `safePtr` and `safeUnwrapPtr`.

jonathanknowles committed:
Add statement of properties to `Cardano.Ledger.Credential.Safe`.

sevanspowell committed:
Fix haskell-language-server build

Anviking committed:
Update pingPong_2/golden file: fee seems to be smaller

This /probably/ is because the default cost model we use in the test
setup has changed to Vasil values with this commit:

https://github.com/input-output-hk/plutus/commit/e9ff3177396f60a0315dec7c525031588b6a7ec6

balanceTx integration tests still work in alonzo.

Anviking committed:
Bump cardano-node, use aeson 2, and use cabal as source of truth

Co-Authored-By: Hamish Mackenzie <Hamish.K.Mackenzie@gmail.com>
Co-Authored-By: Evgenii Akentev <i@ak3n.com>
Co-Authored-By: Samuel Evans-Powell <mail@sevanspowell.net>
Co-Authored-By: Johannes Lund <johannes.lund@iohk.io>

Squashed from:

- Update hackageNix + flake
- Bump iohk-nix
- Initial attempts at bumping dependencies
- Ensure we're using ghc-8.10.7
- Try using cabal.project as source of truth
- Explore alternative path: tweak existing LTS
- Make `stackProject` and `cabalProject` work the same.
- Remove stack-nix materialization
    This fixes this error from running ./nix/regenerate.sh:

    ```
    error: attribute 'stack-nix' missing

           at /nix/store/flpvyc5zbczv3zmspqylpmpmq2sh3wwq-source/flake.nix:371:24:

              370|               inherit (pkgs) sha256map-regenerate checkStackProject;
              371|               inherit (project.stack-nix.passthru) generateMaterialized;
                 |                        ^
              372|               buildToolsGenerateMaterialized = pkgs.haskell-build-tools.regenerateMaterialized;
    ```

    Which we need to be able to run to bump dependencies (ran into this when
    bumping cardano-addresses in another, to-be future commit)

- Tweak MonadSTM instance
- Update to aeson2 and partially to recent cardano-node
- Bump cardano-addresses
- Partially bump node again
- Start filling in some Babbage glue
- Fix cardano-wallet-core haddock comments
- Fix genTxIx
  Fixes several test failures of type

      Transaction input index is too big, acceptable value is up to 2^32-1, in input TxIn "01f4b788593d4f70de2a45c2e1e87088bfbdfa29577ae1b62aba60e095e3ab53" (TxIx 18446744073709551615)
- fix warnings
- Make TPraos type parameters concrete
- fixups

Anviking committed:
rebase fixup: remove redundant Small import in Cardano.Api.Gen

Anviking committed:
rebase fixup: aeson 2 changes in Blockfrost module

Anviking committed:
tmp: disable nix regeneration of sha256map

Anviking committed:
tweak formatting

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Anviking committed:
fix hlint

Anviking committed:
Try updating haskell.nix

Anviking committed:
Try to fix nix `error: attribute 'defaultPackage' missing`

related to emanote. Looking with `nix repl`, `:lf .` it seems the
attribute was renamed/moved and that this commit should work...
hopefully...

Anviking committed:
tmp: try removing libsodium-vrf addition from nix

sevanspowell committed:
haskell.nix changes

Anviking committed:
try setting reinstallableLibGhc to true

Anviking committed:
Remove explicit emanote input from flake.nix

In response to:

    This package indirectly depends on multiple versions of the same package. This is very likely to cause a compile failure.
      package unionmount (unionmount-0.2.0.0-1c0VW38SduEFwyZpC0D9mI) requires with-utf8-1.0.2.3-1X9YYofxvRA1CtBYRjouBd
      package emanote (emanote-0.6.11.1) requires with-utf8-1.0.2.3-1X9YYofxvRA1CtBYRjouBd
      package emanote (emanote-0.6.11.1) requires with-utf8-1.0.2.3-1X9YYofxvRA1CtBYRjouBd
      package tailwind (tailwind-0.3.0.0-ApvnIqbkjPPF9kmRqd9nHW) requires with-utf8-1.0.2.3-8SBXNbv7SJCLUwaAiVprV7

It seemed to make `:b hydraJobs.macos.intel.shells.docs` build locally;
I hope it makes CI greener too.

Anviking committed:
Regenerate cli goldens (only whitespace changes)

Anviking committed:
Replace nix/sha256map and stack.yaml with `checkCabalProject`

sevanspowell committed:
Add -fexternal-interpreter GHC option to plutus-core when profiling

- The plutus repositorys (CONTRIBUTING
file)[https://github.com/input-output-hk/plutus/blob/master/CONTRIBUTING.adoc#how-to-build-the-code-with-profiling]
helpfully points out that the -fexternal-interpreter flag is required when
profiling.
- Add this flag to the plutus-core package when profiling is enabled.

Anviking commented on cabal.project:
```diff
@@ -222,25 +234,52 @@ source-repository-package
+     type: git
+     location: https://github.com/haskell-works/hw-aeson
+     tag: d99d2f3e39a287607418ae605b132a3deb2b753f
+     --sha256: 1vxqcwjg9q37wbwi27y9ba5163lzfz51f1swbi0rp681yg63zvn4
```
Probably not needed now, will attempt removing.

Anviking commented on cabal.project:
```diff
@@ -252,33 +291,49 @@ allow-newer:
+package plutus-core
+  ghc-options: -fexternal-interpreter
+  -- Needed to fix https://hydra.iohk.io/build/15694577/nixlog/2/tail
+  -- Cf. https://input-output-rnd.slack.com/archives/C21UF2WVC/p1624564618267800?thread_ts=1624563615.267000&cid=C21UF2WVC
```
Will try reverting in case it wasn't necessary.

Anviking commented on cabal.project:
```diff
@@ -252,33 +291,49 @@ allow-newer:
+  flags: +systemd
 package cardano-node
-  flags: -systemd
+  flags: +systemd
```
🧐 Not sure where this came from nor the implications of it. Do we want it?

edit: it's from https://github.com/input-output-hk/cardano-wallet/pull/3293

Anviking replied:
I see the node repo has docs which says

> We recommend [systemd](https://www.freedesktop.org/wiki/Software/systemd/) on Linux

and has  `systemd = !pkgs.stdenv.hostPlatform.isMusl` in its nix config.

With this change we have `flags: +systemd` but which can be overridden to [false if isMusl](https://github.com/input-output-hk/cardano-wallet/blob/1f8742817b9c7a2ab4a819d3ed284931880c371d/nix/haskell.nix#L407-L409), which should be equivalent to the node's `= !isMusl`.

So while I don't grasp the implications, I'd guess this is fine.

Anviking commented on lib/core/cardano-wallet-core.cabal:
```diff
@@ -419,6 +421,9 @@ test-suite unit
     , warp
+  if (flag(scrypt))
+    cpp-options: -DHAVE_SCRYPT
+    build-depends: scrypt
```
Ah, yes. The unit tests were still unconditionally dependent on `scrypt`, unlike the library, so this seems good. Haven't checked whether it passes, but this seems fair nonetheless.

Anviking committed:
Remove now redundant dependency on hw-aeson

Anviking commented on lib/core/src/Cardano/Api/Gen.hs:
```diff
@@ -266,7 +269,26 @@ genTxInsCollateral era =
                           ]
 
 genSlotNo :: Gen SlotNo
-genSlotNo = SlotNo <$> arbitrary
+genSlotNo = do
+    boundary <- genBoundary
+    frequency
+        [ (20, pure $ SlotNo boundary)
+        , (20, pure $ SlotNo (maxBound @Word64 - boundary) )
+        , (60, SlotNo <$> arbitrary @Word64)
+        ]
+  where
+    genBoundary = choose (0, 10_000)
+
+genSlotNo32 :: Gen SlotNo32
+genSlotNo32 = do
+    offset <- genOffset
+    frequency
+        [ (20, pure $ SlotNo32 offset)
+        , (20, pure $ SlotNo32 (maxBound @Word32 - offset) )
+        , (60, SlotNo32 <$> arbitrary @Word32)
+        ]
+  where
+    genOffset = choose (0, 10_000)
```
👍

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1210,23 +1291,55 @@ _assignScriptRedeemers (toAlonzoPParams -> pparams) ti resolveInput redeemers tx
             Left (UnknownTxIns ins) ->
                 Left $ ErrAssignRedeemersUnresolvedTxIns $ map fromShelleyTxIn (F.toList ins)
+            Left (BadTranslation translationError) ->
+                Left $ ErrAssignRedeemersTranslationError $ show translationError
```
👍

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1185,23 +1247,42 @@ _assignScriptRedeemers (toAlonzoPParams -> pparams) ti resolveInput redeemers tx
 
-        res <- evaluateTransactionExecutionUnits
-                pparams
+        let res = evaluateTransactionExecutionUnits
```
🧐 I think the `ErrAssignRedeemersPastHorizon` gets lost here. (`toEpochInfo` above won't throw it, which is a flaw of `TimeInterpreter`, but still, whether we get a PastHorizon or not would only be known when `evaluateTransactionExecutionUnits` actually tries to convert the validity interval.)

Will try to fix.

Anviking replied:
Ah, it seems we no longer get the actual `PastHorizonException` from the ledger 😬 https://github.com/input-output-hk/cardano-ledger/blob/e290bf8d0ea272a51e9acd10adc96b4e12e00d37/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs#L106

Anviking replied:
We probably should remove `ErrAssignRedeemersPastHorizon` then, and expose each constructor of `TranslationError` separately in the our API.

Pushed this for now: 3c62f15f3b7879b4e0460ab20647b0280e0a9d4c

Anviking replied:
Or perhaps we can minimise the api-breakage by still using the `past_horizon` error code but with less details than the full `ErrAssignRedeemersPastHorizon`. Will think/look into tomorrow.

Anviking replied:
I pushed 896c0680766b77fc761bdf0313d8ca25977bcbdb to minimise the API breakage from this PR and keep using the `past_horizon` error code. cc @jonathanknowles does it sound somewhat reasonable?

Anviking committed:
Remove unused ErrAssignRedeemersPastHorizon for now

It seems we can no longer construct this error as the ledger no longer
returns the actual `PastHorizon`. This commit removes it, with the
intention that we can re-add it being explicitly rendered in our API
instead of being lumped under TranslationErrors. It would probably also
be nice to render all the other TranslationErrors explicitly.

Anviking committed:
Re-format a few lines which are too long

Anviking committed:
Apply suggestion: use \case in shelleyBasedToCardanoEra

sevanspowell committed:
hie-bios: Set reinstallableLibGhc to false

paweljakubas reviewed (approved):
Let's merge this beast! ;-)

Anviking committed:
Workaround hie-direnv.yaml errors

Maybe a haskell.nix regression.

https://input-output-rnd.slack.com/archives/CAP8NM7N0/p1655271607028309

jonathanknowles commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -54,7 +55,7 @@ import Data.Text
 import Data.Text.Class
     ( FromText (..), TextDecodingError (..), toText )
-import Options.Applicative
+import "optparse-applicative" Options.Applicative
```
This is the first time I've seen us use the `PackageImports` extension. It's very lucky that this extension exists! Though it's also frustrating that it has to exist.

Anviking replied:
For the record, Sam elaborated further in: [604f06e](https://github.com/input-output-hk/cardano-wallet/pull/3280/commits/604f06e1fe02cd69bb119cdbc068237a4617ddc8)

jonathanknowles committed:
Miscellanous formatting fixes.

jonathanknowles commented on flake.nix:
```diff
@@ -33,7 +33,7 @@
 
   inputs = {
-    nixpkgs.follows = "haskellNix/nixpkgs-2111";
+    nixpkgs.follows = "haskellNix/nixpkgs-unstable";
```
Do we know what the implications of following `nixpkgs-unstable` are?

For example, will it lead to more frequent fetching of dependencies? (i.e., when entering a `nix develop` shell, will we have to download dependencies more frequently than before?)

Anviking replied:
> Do we know what the implications of following nixpkgs-unstable are?

I don't entirely, but it seems to be needed, and I believe the consensus is that it's "fine".

>  (i.e., when entering a nix develop shell, will we have to download dependencies more frequently than before?)

No, the dependencies are locked in `flake.lock`, and we'd only need to download new dependencies when we bump... `haskell.nix` I think.

jonathanknowles commented on lib/core/src/Cardano/Api/Gen.hs:
```diff
@@ -1284,7 +1323,10 @@ genTxBodyContent era = do
+            , Api.txInsReference = TxInsReferenceNone
             , Api.txOuts
+            , Api.txTotalCollateral = TxTotalCollateralNone
+            , Api.txReturnCollateral = TxReturnCollateralNone
```
@Anviking @paweljakubas 

We generate a value for `txIns` here. Do we need a task to generate values for `txInsCollateral`, `txTotalCollateral`, and `txReturnCollateral`, and `txInsReference`?

Anviking replied:
Hm:

- `txTotalCollateral` - might be added as part of ADP-1653
- `txReturnCollateral` - might be added as part of ADP-1653
- `txInsReference` - relevant to ADP-1655
- `txInsCollateral` - <s>I have added it to https://input-output.atlassian.net/browse/ADP-1374</s>

Anviking replied:
I don't think their addition should be _strictly_ required in ADP-1653 or ADP-1655 respectively, and at the same time I suspect creating a new ticket at the moment might just add noise.

Anviking replied:
0902c88eaefb7bc833134664a802e4a8ce49cc52

And actually we are already generating `txInsCollateral`, just at a later step (added a note).

jonathanknowles committed:
Explicitly import constructors for `BasicFailure`.

jonathanknowles committed:
Use `TranslationError` within `ErrAssignRedeemersTranslationError`.

We can avoid a "stringly-typed" constructor here by reusing the
`TranslationError` type provided by `Cardano.Ledger.Alonzo.TxInfo`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -507,10 +506,10 @@ data ErrAssignRedeemers
     | ErrAssignRedeemersUnresolvedTxIns [TxIn]
     -- ^ The transaction contains inputs which couldn't be resolved.
+    | ErrAssignRedeemersTranslationError String
+    -- ^ Mistranslating of hashes, credentials, certificates etc.
```
@Anviking Perhaps we can avoid this "stringly-typed" constructor by reusing `Cardano.Ledger.Alonzo.TxInfo.TranslationError` here?

We can defer the conversion to `String` to the `IsServerError` instance for `ErrAssignRedeemers`.

I've added 035fd118f60bb9b0b79b10fc7988347849649fce. Though feel free to remove if you feel it is inappropriate.

Anviking replied:
Thanks, that's better!

jonathanknowles reviewed (approved):
This looks reasonable to me.

I just have a few questions! (See comments.)

sevanspowell committed:
Link usage of PackageImports to ticket explaining reasoning

- Link usages of PackageImports to ADP-1910, where we explain why we used
  PackageImports and how we might get rid of it in future.

sevanspowell committed:
Try bumping haskell.nix

- Try bumping haskell.nix to fix remaining issues with windows cross-compile
builds.

Anviking committed:
Update node version in readme compatibility matrix

Anviking committed:
Improve TODOs and linking to tasks in Api.Gen module

Anviking committed:
Minimize API changes w.r.t `TimeTranslationPastHorizon`

By treating `TimeTranslationPastHorizon` differently we can return the
same api error code as we used to, and which is arguably more
descriptive.

sevanspowell committed:
Merge pull request #3328 from input-output-hk/anviking/ADP-1744/completely-drop-stack.yaml

Completely drop stack.yaml and sha256map.nix
